### ISSUE TITLE
Clarification on watch when server (or store) is partitioned

### ISSUE THREAD

**Issue by embano1** [kubernetes/kubernetes#89488]
After reading up on some corner cases with etcd client behavior (connections, streams, etc.) I came across the following option that has been introduced in etcd a while ago:

```go
// WithRequireLeader requires client requests to only succeed
// when the cluster has a leader.
func WithRequireLeader(ctx context.Context) context.Context {...}
```
[(Source)](https://github.com/etcd-io/etcd/blob/67da93f739e33ee9b34792fccba2d0c100264ea4/clientv3/ctx.go#L28)

From the official etcd docs on client behavior:

> Client-side keepalive ping still does not reason about network partitions. Streaming request may get stuck with a partitioned node. Advanced health checking service need to be implemented to understand the cluster membership (see etcd#8673 for more detail).
Source: [clientv3-grpc1.23: Balancer Limitation](https://etcd.io/docs/v3.4.0/learning/design-client/)

I do not see this option being used in the API server when `Watch()` is [established](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L218).

Is there code in the API server that deals with these cases (etcd/API server node paritioned) and disconnects clients (REST, SDK) appropriately? Or are we at risk of having dangling (hanging) consumer-side client watches due to not using this option in the etcd client?

Related discussions:

[clientv3: clarify "WithRequireLeader" for network partition](https://github.com/etcd-io/etcd/pull/9872)
[Network-partition aware health service](https://github.com/etcd-io/etcd/issues/8673)

/sig api-machinery

cc/ @dims @jingyih @jpbetz @sttts

**Comment by dims** [comment]
@jpbetz was this something that was considered already?

**Comment by jpbetz** [comment]
We should be using `WithRequireLeader` for watches. (see also the [etcd Watcher godoc](https://github.com/etcd-io/etcd/blob/3ac7a1151519af0816647d5f0757119c3f4a24ac/clientv3/watch.go#L65)).

It should just need to be added [here](https://github.com/kubernetes/kubernetes/blob/5d4553a90d110391d20324fe7f1890e48c8de659/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go#L214).

It would also be valuable to confirm that the behavior works as expected both when establishing watches and when a etcd member is partitioned after the watch has been established.

**Comment by jpbetz** [comment]
If anyone wants to own fixing this please self assign, else feel free to assign to me.

**Comment by tedyu** [comment]
For this method:
```
func WithRequireLeader(ctx context.Context) context.Context {
```
It seems we should do:
```
diff --git a/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go b/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go
index ca1583fb379..1bc841ca653 100644
--- a/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go
@@ -126,7 +126,7 @@ func (w *watcher) createWatchChan(ctx context.Context, key string, rev int64, re
                // The filter doesn't filter out any object.
                wc.internalPred = storage.Everything
        }
-       wc.ctx, wc.cancel = context.WithCancel(ctx)
+       wc.ctx, wc.cancel = context.WithCancel(clientv3.WithRequireLeader(ctx))
        return wc
 }
```

**Comment by jingyih** [comment]
Thanks! This is a good question. I am not exactly sure what is the semantics on kubernetes watch. It is correct that without `WithRequireLeader`, etcd server will not cancel watch stream when itself is isolated from the cluster. However, when this happens, other requests to etcd (including health checks on etcd) will fail. So the control plane is already in a bad state and is likely to be replaced / repaired either automatically or manually. IIUC, this means apiserver is will restart, the its clients will restart their watches.

**Comment by jingyih** [comment]
Oh I just saw Joe's earlier comment. If using `WithRequireLeader` aligns with k8s watch semantics, we should use it:)

**Comment by dims** [comment]
@jingyih i'd support moving to 3.4.5 to pick this up!

**Comment by embano1** [comment]
@jpbetz I am happy to contribute but want to do some more testing and validation on the current behavior of `Watch` and with the proposed fix. Please give me some days.

**Comment by embano1** [comment]
Quick update: after some quick tests it seems that `WithRequireLeader()` does not solve the issue with an active `Watch()` after the connected etcd endpoint becomes partitioned.

I think we have the following scenarios to consider:

1) Single node API server/etcd
2) Quorum-based cluster (e.g. 3 nodes) where K8s control plane **is colocated** with the etcd nodes (e.g. one API server and one etcd instance on one node)
3) Quorum-based cluster (e.g. 3 nodes) where K8s control plane **is not colocated** with the etcd endpoints (separate nodes for K8s control plane and etcd nodes)

We can rule out **(1)** because technically a partition with just one node in the cluster (K8s control plane and etcd) cannot happen.

**(3)** is mostly addressed with existing gRPC timeouts and keep-alives so when an API server connected to an etcd node that gets partitioned this API server will reconnect to other reachable etcd nodes (endpoints). However, this is only true if the API server itself is not within that partition (e.g. multi-zone deployment where a whole zone becomes partitioned; read on for my observations).

**(2)** is the interesting one: an API server is connected to an etcd node and both become partitioned from the majority. But the API server has active watches against this etcd node and serves these (Kubernetes) watches to Kubernetes clients (controllers, etc.). I was expecting that the `WithRequireLeader()` gets evaluated even after establishing the `Watch()`. But it seems my assumption is not correct, i.e. it is only evaluated (checked) when `Watch()` is called. At least I did not see an existing `Watch()` being dropped after partitioning the nodes from the quorum (I did check the underlying gRPC transport status and it remained in `"CONNECTING"` stayed for as long as the partition was not resolved.

So what does that mean for the issue raised here? One could argue that if more than just one etcd node is partitioned, i.e. also the API server and perhaps other Kubernetes controllers running in that partitioned environment (e.g. a zone), it makes no difference using `WithRequireLeader()` because:

- the clients (custom controllers, etc.) are also partitioned and thus cannot connect to any other API server anyways; would be nice to return an error here instead of hanging but the system won't self-heal during the partition
- even if the clients (custom controllers) would not be in that partition, the corresponding API server could not be reached anymore and client-go session handling should address this (e.g. reconnect to another API server)

@jingyih you write:

> It is correct that without WithRequireLeader, etcd server will not cancel watch stream when itself is isolated from the cluster

I did not see an existing watch being dropped, e.g. using this simple `clientv3` code:

```go
	leaderCtx := clientv3.WithRequireLeader(ctx)
	// rch := cli.Watch(context.Background(), "", clientv3.WithPrefix(), clientv3.WithRev(0+1), clientv3.WithPrevKV())

	fmt.Println("starting watch")
	rch := cli.Watch(leaderCtx, "", clientv3.WithPrefix(), clientv3.WithRev(rev+1))
	for wresp := range rch {
		...
		}
	}
```

Once I partition the etcd node including this client it will not drop the `Watch()`. Looking at the definition of `WithRequireLeader()` it seems that this is performed only during establishing the `Watch()` but not reevaluated as long as the client has a successful connection to the etcd node (e.g. node is partitioned but not dead).

```go
// WithRequireLeader requires client requests to only succeed
// when the cluster has a leader.
func WithRequireLeader(ctx context.Context) context.Context {
	md, ok := metadata.FromOutgoingContext(ctx)
	if !ok { // no outgoing metadata ctx key, create one
		md = metadata.Pairs(rpctypes.MetadataRequireLeaderKey, rpctypes.MetadataHasLeader)
		return metadata.NewOutgoingContext(ctx, md)
	}
	copied := md.Copy() // avoid racey updates
	// overwrite/add 'hasleader' key/value
	metadataSet(copied, rpctypes.MetadataRequireLeaderKey, rpctypes.MetadataHasLeader)
	return metadata.NewOutgoingContext(ctx, copied)
}
```
https://github.com/etcd-io/etcd/blob/67da93f739e33ee9b34792fccba2d0c100264ea4/clientv3/ctx.go#L26

Since I am by no means an expert on the etcd `Watch()` semantics here, I might be doing something wrong. It would be great to get some clarification whether this issue (I raised, I know :) ...) needs to be addressed, documented, or can actually be addressed at all with the current implementation of `WithRequireLeader()`.

**Comment by dims** [comment]
@jpbetz @xiang90 @jingyih @gyuho Please see analysis above. should this work when the etcd + apiserver are on the same node or not?

**Comment by jpbetz** [comment]
Thanks for the careful analysis @embano1 

My take on this is:

For case **(2)** where the apiserver and etcd are co-located (and the apiserver only talks to that etcd), I don't see any reason why dropping the watch would help. In this type of setup, when etcd becomes partitioned, the co-located apiserver becomes unavailable and apiserver clients should route away from that apiserver entirely until the partition is healed. Even if the apiserver were to drop the watch, it is still unavailable, and the only etcd it can 

For case **(3)** the behavior is already better than I expected. `WithRequireLeader` optimizes the case slightly by ensuring the initial watch connection is against a non-partitioned etcd, so it's still worth enabling, but not urgently needed since failover still happens.

We should document in etcd that `WithRequireLeader` only applies to establishing the initial watch connection, the documentation is currently misleading.

**Comment by jingyih** [comment]
> @jingyih you write:
It is correct that without WithRequireLeader, etcd server will not cancel watch stream when itself is isolated from the cluster

This is correct. Please refer to the following test which verifies this behavior. The server should close the watch channel when itself is isolated from the cluster.
https://github.com/etcd-io/etcd/blob/a04b74d4b4ebced2aca96dde08cb73b04bcdc484/clientv3/integration/watch_test.go#L789

> Looking at the definition of WithRequireLeader() it seems that this is performed only during establishing the Watch() but not reevaluated as long as the client has a successful connection to the etcd node (e.g. node is partitioned but not dead).

IIUC, this is not correct. When creating the watch request, if the context has the metadata written by `WithRequireLeader `, etcd server will keep monitor its leader for this watch stream.
https://github.com/etcd-io/etcd/blob/a04b74d4b4ebced2aca96dde08cb73b04bcdc484/etcdserver/api/v3rpc/interceptor.go#L232

**Comment by jpbetz** [comment]
Thanks @jingyih, that means the documentation is correct then.

So while we haven't found any pressing issues that `WithRequireLeader` would fix, it does do what it should do, enabling it in the kube-apiserver won't make anything worse, and might make failover slightly faster?

**Comment by jingyih** [comment]
> enabling it in the kube-apiserver won't make anything worse, and might make failover slightly faster?

Make sense to me.

**Comment by jingyih** [comment]
BTW, I just did a quick test. 

1. start an etcd cluster of 3 members: A, B and C.
2. run `etcdctl watch foo --endpoints={endpoint of member A}`
3. stop member B and C
4. after some time, the etcdctl command prints `watch was canceled (etcdserver: no leader)`

**Comment by embano1** [comment]
Sorry for my delayed reply, wanted to do some more tests before replying.

@jingyih 

> BTW, I just did a quick test.

I can confirm this scenario. **However**, the tests that I did and which caught my attention because it was not what I expected from reading the doc string was to not shut down the peers but **isolate one peer**, i.e. simulate a partition which technically is also a "lost leader" scenario.

In this case I saw the watches not being closed (neither for `etcdctl` nor programmatically). Not sure if this is a bug or by design but I think it's more likely to have partitions than loosing the whole cluster where a majority of nodes is terminated/crashed (stop-failure). 

If you think it's worth and not expected behavior I am happy to open an issue in the etcd repo and provide more details.

**Comment by jpbetz** [comment]
@embano1 I'm in favor of you opening a new against etcd for the watch not being closed in the lost leader scenario (and linking to this discussion). If it turns out it is a duplicate we can always mark as such.

**Comment by gyuho** [comment]
https://godoc.org/github.com/etcd-io/etcd/clientv3#Watcher has some detailed watch behaviors.

```
type Watcher interface {
    // Watch watches on a key or prefix. The watched events will be returned
    // through the returned channel. If revisions waiting to be sent over the
    // watch are compacted, then the watch will be canceled by the server, the
    // client will post a compacted error watch response, and the channel will close.
    // If the requested revision is 0 or unspecified, the returned channel will
    // return watch events that happen after the server receives the watch request.
    // If the context "ctx" is canceled or timed out, returned "WatchChan" is closed,
    // and "WatchResponse" from this closed channel has zero events and nil "Err()".
    // The context "ctx" MUST be canceled, as soon as watcher is no longer being used,
    // to release the associated resources.
    //
    // If the context is "context.Background/TODO", returned "WatchChan" will
    // not be closed and block until event is triggered, except when server
    // returns a non-recoverable error (e.g. ErrCompacted).
    // For example, when context passed with "WithRequireLeader" and the
    // connected server has no leader (e.g. due to network partition),
    // error "etcdserver: no leader" (ErrNoLeader) will be returned,
    // and then "WatchChan" is closed with non-nil "Err()".
    // In order to prevent a watch stream being stuck in a partitioned node,
    // make sure to wrap context with "WithRequireLeader".
    //
    // Otherwise, as long as the context has not been canceled or timed out,
    // watch will retry on other recoverable errors forever until reconnected.
    //
    // TODO: explicitly set context error in the last "WatchResponse" message and close channel?
    // Currently, client contexts are overwritten with "valCtx" that never closes.
    // TODO(v3.4): configure watch retry policy, limit maximum retry number
    // (see https://github.com/etcd-io/etcd/issues/8980)
    Watch(ctx context.Context, key string, opts ...OpOption) WatchChan
```

```
    // In order to prevent a watch stream being stuck in a partitioned node,
    // make sure to wrap context with "WithRequireLeader".
```

**Comment by gyuho** [comment]
FYI. There was a bug in `WithRequireLeader`. The fix was backported to >= v3.3.19 and v3.4.5 https://github.com/etcd-io/etcd/pull/11687.

**Comment by embano1** [comment]
Thank you @gyuho and @jpbetz for your feedback.
After reviewing the docs and https://github.com/etcd-io/etcd/issues/8980 I think the issue with `Watch()` and how to deal with partitions is still open (see TODO below).

```go
// TODO(v3.4): configure watch retry policy, limit maximum retry number
// (see https://github.com/coreos/etcd/issues/8980)
```

The scenario where clients (Watchers or any client indirectly consuming from an etcd `Watch()`) would hang (forever?) is when the client gets partitioned with the corresponding partitioned etcd node **OR** when a whole zone/dc is partitioned. Since etcd picks safety (consistency) over liveness (availability) in this case at least not data corruption occurs but the overall system would be stuck when using watches instead of PUT/GET RPCs with timeouts. 

It would be ideal to throw a useful error, e.g. "lost leader", instead of hanging in "gRPC TRANSIENT_ERROR" state (which is what I'm seeing). Until this is resolved in etcd, perhaps the docs could be updated as `WithRequireLeader` is a bit misleading here.

I will comment on the etcd issue above and file a PR for the K8s watcher implementation to make use of `WithRequireLeader` as it is still a useful option when (re)establishing the stream. Depending on the environment, API server instances could have a health/liveness probe (or systemd equivalent) which could restart the process in case of a partition. Upon reconnect to the etcd node with the `WithRequireLeader` option it would immediately fail and error out accordingly.
